bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+}
+
+/// Free a set of bytes. The first parameter can be null in cases where a
+/// tmprl_core_t instance isn't available. If the second parameter is null, this
+/// is a no-op.
+#[no_mangle]
+pub extern "C" fn tmprl_bytes_free(core: *mut tmprl_core_t, bytes: *const tmprl_bytes_t) {
+    // Bail if freeing is disabled
+    unsafe {
+        if bytes.is_null() || (*bytes).disable_free {
+            return;
+        }
+    }
+    let bytes = bytes as *mut tmprl_bytes_t;
+    // Return vec back to core before dropping bytes
+    let vec = unsafe { Vec::from_raw_parts((*bytes).bytes as *mut u8, (*bytes).len, (*bytes).cap) };
+    // Set to null so the byte dropper doesn't try to free it
+    unsafe { (*bytes).bytes = std::ptr::null_mut() };
+    // Return only if core is non-null
+    if !core.is_null() {
+        let core = unsafe { &mut *core };
+        core.return_buf(vec);
+    }
+    unsafe {
+        Box::from_raw(bytes);
+    }
+}
+
+// Used for maintaining pointer to user data across threads
+struct UserDataHandle(*mut libc::c_void);
+unsafe impl Send for UserDataHandle {}
+unsafe impl Sync for UserDataHandle {}
+
+lazy_static::lazy_static! {
+    static ref DEFAULT_INIT_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::InitResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_SHUTDOWN_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::ShutdownResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_REGISTER_WORKER_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::RegisterWorkerResponse::default().encode_to_vec())
+    };
+}
+
+/// A runtime owned by Core. This must be passed to tmprl_runtime_free when no
+/// longer in use. This must not be freed until every call to every tmprl_core_t
+/// instance created with this runtime has been shutdown.
+pub struct tmprl_runtime_t {
+    tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
+}
+
+/// Create a new runtime. The result is never null and must be freed via
+/// tmprl_runtime_free when no longer in use.
+#[no_mangle]
+pub extern "C" fn tmprl_runtime_new() -> *mut tmprl_runtime_t {
+    Box::into_raw(Box::new(tmprl_runtime_t {
+        // TODO(cretz): Options to configure thread pool?
+        tokio_runtime: std::sync::Arc::new(
+            tokio::runtime::Builder::new_multi_thread()
+                .enable_all()
+                .build()
+                .unwrap(),
+        ),
+    }))
+}
+
+/// Free a previously created runtime.
+#[no_mangle]
+pub extern "C" fn tmprl_runtime_free(runtime: *mut tmprl_runtime_t) {
+    if !runtime.is_null() {
+        unsafe {
+            Box::from_raw(runtime);
+        }
+    }
+}
+
+/// A core instance owned by Core. This must be passed to tmprl_core_shutdown
+/// when no longer in use which will free the resources.
+pub struct tmprl_core_t {
+    tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
```

Comment:
Why does this guy have it's own runtime instead of the `tmprl_runtime_t`?

Additionally it probably makes sense to have this type be an alias for an `Arc` around a struct with these fields rather than using two `Arc`s, if both fields are needed.

Reply:
> Why does this guy have it's own runtime instead of the tmprl_runtime_t?

They're shared instances of the same runtime. I just felt this read simpler than embedding the runtime struct in here. Technically this is a bit safer because on `tmprl_runtime_free`, that will just decrement a reference count instead of segfault here.

Reply:
Segfault could be prevented w/ custom drop impl, but yeah I get it. 

I would make a field docstring indicating that it's always going to be the same runtime as whatever was in the runtime struct.

It also sorta makes me wonder if it makes sense to expect the runtime to be passed by ownership into `tmprl_core_init`, and then free it on core shutdown finishing. Do we have a reason for C/lang to want to keep that runtime around independently of core? If it wants to do other async-y things, perhaps it makes sense to just use a different runtime for that anyway, and Core always operates on its own dedicated runtime?

Reply:
> I would make a field docstring indicating that it's always going to be the same runtime as whatever was in the runtime struct.

:+1:

> Do we have a reason for C/lang to want to keep that runtime around independently of core? If it wants to do other async-y things, perhaps it makes sense to just use a different runtime for that anyway, and Core always operates on its own dedicated runtime?

I think Tokio's task management and scheduling and thread pool works better if reused. Say you had 10 core instances because you were running on 10 namespaces. A Tokio thread pool shared across all of them in theory may be better performing than separate ones.

By default Tokio will make a thread pool of core count for each runtime. I could also see this being configurable (or other things like thread name) in the future and the separation of runtime and core to the caller is logical since it's the separation that exists.